aten/src/ATen/core/boxing/KernelFunction.h

```diff
@@ -0,0 +1,397 @@
+#pragma once
+
+#include <ATen/core/stack.h>
+#include <c10/util/TypeList.h>
+#include <ATen/core/boxing/kernel_functor.h>
+#include <ATen/core/boxing/kernel_function.h>
+#include <ATen/core/boxing/kernel_lambda.h>
+
+namespace c10 {
+
+namespace detail {
+template<class Return, class... Args> struct boxAndCallBoxedFunc;
+}
+
+/**
+ * KernelFunction is similar to std::function but stores a kernel function.
+ * You can create a KernelFunction from a boxed or unboxed function/functor/lambda
+ * and call it in a boxed or unboxed way. If the way it was created doesn't
+ * match the way it was called, it will do boxing or unboxing as necessary.
+ */
+class CAFFE2_API KernelFunction final {
+public:
+  using BoxedKernelFunction = void(OperatorKernel*, Stack*);
+
+  KernelFunction()
+  : functorFactory_()
+  , functor_(nullptr)
+  , boxed_kernel_func_(nullptr)
+  , unboxed_kernel_func_(nullptr)
+  {}
+
+  bool isValid() const {
+    // TODO We want to introduce the invariant that all kernels must be callable in a boxed way, then this should only check boxed_kernel_func_.
+    return boxed_kernel_func_ != nullptr || unboxed_kernel_func_ != nullptr;
+  }
+
+  /**
+   * Call the function in a boxed way.
+   * If the kernel function was created with an unboxed function,
+   * this will call an unboxing wrapper which then calls into that
+   * unboxed function.
+   *
+   * Example:
+   *
+   * > void boxed_func(OperatorKernel*, Stack* stack) {...}
+   * > KernelFunction func = KernelFunction::makeFromBoxedFunction(&boxed_func);
+   * > Tensor result = func.callBoxed(stack);
+   *
+   * Or, with an unboxed implementation:
+   *
+   * > KernelFunction func = KernelFunction::makeFromUnboxedLambda(
+   * >      [] (Tensor a, bool b) -> Tensor {...});
+   * > Tensor result = func.callBoxed(stack);
+   */
+  void callBoxed(Stack* stack) const {
+    if (C10_UNLIKELY(boxed_kernel_func_ == nullptr)) {
+      if (unboxed_kernel_func_ == nullptr) {
+        TORCH_INTERNAL_ASSERT(false, "Tried to call KernelFunction::callBoxed() on an uninitialized KernelFunction.");
+      } else {
+        // TODO We want to introduce the invariant that all kernels must be callable in a boxed way, then this case should be impossible.
+        TORCH_INTERNAL_ASSERT(false, "Tried to call KernelFunction::callBoxed() on a KernelFunction that can only be called with KernelFunction::callUnboxed().");
+      }
+    }
+
+    (*boxed_kernel_func_)(getFunctor_(), stack);
+  }
+
+  /**
+   * Call the function in an unboxed way.
+   * As the "Only" in the name suggests, this only works for KernelFunctions
+   * that are backed by an unboxed kernel. If the KernelFunction was created
+   * in a boxed way, this will fail (also see KernelFunction::callUnboxed()).
+   *
+   * KernelFunction::callUnboxed() is generally better, since it will allow
+   * calling KernelFunctions that are backed by either boxed or unboxed
+   * kernels, but that one will not work for all types.
```

Comment:
I don't understand, is this a compilation-time error? I would prefer if we have a single entry point `callUnboxed`, that just skips the fallback to boxed types if the types in question are not supported. This is how I implemented it in #26368

Reply:
Yes, I chose a bit a different approach here. I ant the  compiler to show an error if you have an op with `use_c10_dispatcher: full` but the boxed fallback doesn't work. That means the `use_c10_dispatcher` flag is the only configuration setting you have to move to fully migrate ops over to the c10 dispatcher. If you have some SFINAE config to remove ops from creating boxed wrappers, that would add an additional configuration setting and make the system more complex.

Reply:
I don't buy the argument that it's more complex. The reason I don't buy it is because #26368 is able to avoid a bunch of pieces which you have to do in this PR:

- There's no per-operator setting saying whether or not it's `full` or `unboxedOnly`. The templates figure it out automatically.
- There's no change to code generation, as is here. The templates figure it out automatically.
- The actual difference in the dispatcher is a single if statement https://github.com/pytorch/pytorch/pull/26368/files#diff-4f2e119446bc2edf856334581e5c70c9R217 testing if the boxed version is supported, and error if it is not

The only downside is that you don't get compile time errors in the way you've done here. But it's easy to turn all of the runtime errors into compile time errors, to see if you have 100% coverage: change `supports_boxed_fallback` to return unconditionally true.